orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -0,0 +1,255 @@
+    ONNX_NAMESPACE::TypeProto type_proto(*(base_arg->TypeAsProto()));
+    return graph.GetOrCreateNodeArg(new_name, &type_proto);
+  };
+  if (send_fw && recv_bw) {
+    // if we have a send forward op followed by a recv backward op, insert WaitEvent and RecordEvent in between.
+    Node* record_node = nullptr;
+
+    // Insert RecordEvent
+    {
+      std::vector<NodeArg*> input_args;
+      std::vector<NodeArg*> output_args;
+      AddInputEvent(graph, "RecordEvent", true /* is_forward */, input_args, new_input_names);
+
+      // Add send forward op's output as record op's input and output
+      for (auto& output : send_fw->MutableOutputDefs()) {
+        auto& new_output = create_node_args(output);
+        output_args.push_back(&new_output);
+        input_args.push_back(output);
+      }
+
+      auto& new_node = graph.AddNode(graph.GenerateNodeName("RecordEvent"),
+                                     "RecordEvent",
+                                     "",
+                                     input_args,
+                                     output_args, /* output */
+                                     {},          /* attribute */
+                                     kMSDomain);
+      record_node = &new_node;
+    }
+    // Insert WaitEvent
+    {
+      std::vector<NodeArg*> input_args;
+      std::vector<NodeArg*> output_args;
+      AddInputEvent(graph, "WaitEvent", false /* is_forward */, input_args, new_input_names);
+
+      input_args.insert(std::end(input_args),
+                        std::begin(record_node->MutableOutputDefs()),
+                        std::end(record_node->MutableOutputDefs()));
+
+      auto& input = recv_bw->MutableInputDefs()[0];
+      auto& new_output = create_node_args(input);
+      output_args.push_back(&new_output);
+      input = &new_output;
+
+      graph.AddNode(graph.GenerateNodeName("WaitEvent"),
+                    "WaitEvent",
+                    "Backward pass",
+                    input_args,
+                    output_args, /* output */
+                    {},          /* attribute */
+                    kMSDomain);
+    }
+  }
+  return Status::OK();
+}
+
+Status TransformGraphForPipeline(Graph& graph) {
+  // insert WaitEvent and RecordEvent to the partition
+  Node* send_fw{nullptr};
+  Node* send_bw{nullptr};
+  Node* recv_fw{nullptr};
+  Node* recv_bw{nullptr};
+  for (auto& node : graph.Nodes()) {
+    if (node.OpType() == "Send") {
+      if (IsBackward(node)) {
+        send_bw = &node;
+      } else {
+        send_fw = &node;
+      }
+    } else if (node.OpType() == "Recv") {
+      if (IsBackward(node)) {
+        recv_bw = &node;
+      } else {
+        recv_fw = &node;
+      }
+    }
+  }
+
+  std::vector<std::string> new_input_names;
+
+  AddRecordBackward(graph, send_bw, new_input_names);
+  ORT_RETURN_IF_ERROR(AddWaitForward(graph, recv_fw, new_input_names));
+  ORT_RETURN_IF_ERROR(AddRecordForwardWaitBackward(graph, send_fw, recv_bw, new_input_names));
```

Comment:
Is `AddRecordForwardWaitBackward` applied to the last pipeline stage? If no, please use an `if` block to avoid calling `AddRecordForwardWaitBackward` for that case.

Reply:
It's apply to all the stages except the last stage :-) The check is down through checking send_fw && recv_bw. It's less ideal. But I think once we have the information of what the current stage is, we can have a better check in this case.

---
In reply to: [409797150](https://github.com/microsoft/onnxruntime/pull/3513#discussion_r409797150) [](ancestors = 409797150)